src/util-lua-ja3.c

```diff
@@ -24,145 +24,183 @@
+#include "app-layer-ssl.h"
+#include "rust.h"
 
-    SSLState *ssl_state = (SSLState *)state;
+static const char ja3_tx[] = "suricata:ja3:tx";
 
-    if (ssl_state->client_connp.ja3_hash == NULL)
-        return LuaCallbackError(luastate, "error: no JA3 hash");
+struct LuaTx {
+    void *tx; // Quic or TLS Transaction
+    AppProto alproto;
+};
 
-    return LuaPushStringBuffer(luastate,
-                               (uint8_t *)ssl_state->client_connp.ja3_hash,
-                               strlen(ssl_state->client_connp.ja3_hash));
+static int LuaJa3GetTx(lua_State *L)
+{
+    AppProto alproto = ALPROTO_QUIC;
+    if (LuaStateNeedProto(L, ALPROTO_TLS)) {
+        alproto = ALPROTO_TLS;
+    } else if (!(LuaStateNeedProto(L, ALPROTO_QUIC))) {
+        return LuaCallbackError(L, "error: protocol nor tls neither quic");
+    }
+    void *tx = LuaStateGetTX(L);
+    if (tx == NULL) {
+        return LuaCallbackError(L, "error: no tx available");
+    }
+    struct LuaTx *ltx = (struct LuaTx *)lua_newuserdata(L, sizeof(*ltx));
+    if (ltx == NULL) {
+        return LuaCallbackError(L, "error: fail to allocate user data");
+    }
+    ltx->tx = tx;
+    ltx->alproto = alproto;
+
+    luaL_getmetatable(L, ja3_tx);
+    lua_setmetatable(L, -2);
+
+    return 1;
 }
 
-static int Ja3GetString(lua_State *luastate)
+static int LuaJa3TxGetHash(lua_State *L)
 {
-    if (!(LuaStateNeedProto(luastate, ALPROTO_TLS)))
-        return LuaCallbackError(luastate, "error: protocol is not tls");
-
-    Flow *f = LuaStateGetFlow(luastate);
-    if (f == NULL)
-        return LuaCallbackError(luastate, "internal error: no flow");
-
-    void *state = FlowGetAppState(f);
-    if (state == NULL)
-        return LuaCallbackError(luastate, "error: no app layer state");
-
-    SSLState *ssl_state = (SSLState *)state;
-
-    if (ssl_state->client_connp.ja3_str == NULL ||
-            ssl_state->client_connp.ja3_str->data == NULL)
-        return LuaCallbackError(luastate, "error: no JA3 str");
-
-    return LuaPushStringBuffer(luastate,
-                               (uint8_t *)ssl_state->client_connp.ja3_str->data,
-                               ssl_state->client_connp.ja3_str->used);
+    struct LuaTx *ltx = luaL_testudata(L, 1, ja3_tx);
+    if (ltx == NULL) {
+        lua_pushnil(L);
+        return 1;
+    }
+    if (ltx->alproto == ALPROTO_TLS) {
+        SSLState *ssl_state = (SSLState *)ltx->tx;
+        if (ssl_state->client_connp.ja3_hash == NULL) {
+            lua_pushnil(L);
+            return 1;
+        }
+        return LuaPushStringBuffer(L, (uint8_t *)ssl_state->client_connp.ja3_hash,
+                strlen(ssl_state->client_connp.ja3_hash));
+    } // else QUIC {
+    const uint8_t *buf = NULL;
+    uint32_t b_len = 0;
+    if (!rs_quic_tx_get_ja3(ltx->tx, STREAM_TOSERVER, &buf, &b_len)) {
+        lua_pushnil(L);
+        return 1;
+    }
+    uint8_t ja3_hash[SC_MD5_HEX_LEN + 1];
```

Comment:
Is it ok to pass a value from the stack ?

Reply:
I believe so, yes. `LuaPushStringBuffer` even does it itself if the input isn't a multiple of 4

Reply:
Thanks, local test was ok, but I was not sure

Reply:
Yeah, Lua ultimately makes a copy of it that it then owns.